transformer_lens/components.py

```diff
@@ -511,6 +543,11 @@ def forward(
         if self.cfg.positional_embedding_type == "rotary":
             q, k = self.rotary_rotate_qk(q, k, kv_cache_pos_offset)
 
+        if self.cfg.dtype not in [torch.float32, torch.float64]:
+            # If using 16 bits, increase the precision to avoid numerical instabilities
+            q = q.to(torch.float32)
+            k = k.to(torch.float32)
+
```

Comment:
I chose to convert to `float32` before the attention layer, instead of dividing sooner with self.attn_scale as proposed. Mostly because it is what is done in the GPT NEO architecture in `def _attn` and it looked a bit more reliable.